src/font/face/coretext.zig

```diff
@@ -400,22 +383,32 @@ pub const Face = struct {
             x += (cell_width - metrics.face_width) / 2;
         }
 
+        // If this is a bitmap glyph, it will always render as full pixels,
+        // not fractional pixels, so we need to quantize its position and
+        // size accordingly to align to full pixels so we get good results.
+        if (sbix) {
+            width = cell_width - @round(cell_width - width - x) - @round(x);
+            height = cell_height - @round(cell_height - height - y) - @round(y);
+            x = @round(x);
+            y = @round(y);
+        }
+
         // Our whole-pixel bearings for the final glyph.
         // The fractional portion will be included in the rasterized position.
-        const px_x: i32 = @intFromFloat(@floor(x));
-        const px_y: i32 = @intFromFloat(@floor(y));
+        const px_x: i32 = @intFromFloat(@floor(x - canvas_padding));
+        const px_y: i32 = @intFromFloat(@floor(y - canvas_padding));
 
         // We keep track of the fractional part of the pixel bearings, which
         // we will add as an offset when rasterizing to make sure we get the
         // correct sub-pixel position.
-        const frac_x = x - @floor(x);
-        const frac_y = y - @floor(y);
+        const frac_x = x - @floor(x - canvas_padding);
+        const frac_y = y - @floor(y - canvas_padding);
 
         // Add the fractional pixel to the width and height and take
         // the ceiling to get a canvas size that will definitely fit
         // our drawn glyph, including the fractional offset.
-        const px_width: u32 = @intFromFloat(@ceil(width + frac_x));
-        const px_height: u32 = @intFromFloat(@ceil(height + frac_y));
+        const px_width: u32 = @intFromFloat(@ceil(width + frac_x + canvas_padding));
+        const px_height: u32 = @intFromFloat(@ceil(height + frac_y + canvas_padding));
```

Comment:
This should be adding `2 * canvas_padding` since the width and height need the padding added to both edges.

Reply:
Fair! I'll include that change in the next push.

Reply:
No, wait a second, the problem is that `canvas_padding` isn't necessarily an integer either, so we'd have to split that too into a rounded and fractional part. That seems more complicated. Maybe we could rename the variables to make the current flow more intuitive?

Reply:
Though if we move synthetic bold pre-constraint as discussed below, the padding is always integral, so this problem goes away.